subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildInclusionCoordinator.java

```diff
@@ -37,12 +42,38 @@
 public class BuildInclusionCoordinator {
     private final Set<IncludedBuildState> loadedBuilds = new CopyOnWriteArraySet<>();
     private final List<IncludedBuildState> libraryBuilds = new CopyOnWriteArrayList<>();
+    private final Map<BuildState, BuildSynchronizer> synchronizers = new ConcurrentHashMap<>();
     private final GlobalDependencySubstitutionRegistry substitutionRegistry;
+    private final WorkerLeaseService workerLeaseService;
     private boolean registerRootSubstitutions;
-    private final Set<BuildState> registering = new HashSet<>();
 
-    public BuildInclusionCoordinator(GlobalDependencySubstitutionRegistry substitutionRegistry) {
+    private static class BuildSynchronizer {
+        final Synchronizer lock;
+        boolean isInProgress;
+
+        BuildSynchronizer(WorkerLeaseService workerLeaseService) {
+            this.lock = workerLeaseService.newResource();
+        }
+
+        void withLock(Runnable action) {
+            lock.withLock(() -> {
+                if (isInProgress) {
```

Comment:
This boolean manipulation exists with a sole purpose - support cycles in vintage mode.